gadopt/stokes_integrators.py

```diff
@@ -352,67 +438,162 @@ def set_solver_options(self) -> None:
+        self.solution_old.assign(self.solution)
+
+
+class StokesSolver(SolverBase):
+    """Solver for the Stokes system.
+
+    Args:
+      solution:
+        Firedrake function representing the field over the mixed Stokes space
+      approximation:
+        G-ADOPT approximation defining terms in the system of equations
+      T:
+        Firedrake function representing the temperature field
+      dt:
+        Float quantifying the time step used in a coupled time integration
+      theta:
+        Float quantifying the implicit contribution in a coupled time integration
+      forcing_term:
+        Firedrake form specifying an additional term contributing to the residual
+      bcs:
+        Dictionary specifying boundary conditions (identifier, type, and value)
+      quad_degree:
+        Integer denoting the quadrature degree
+      solver_parameters:
+        Dictionary of PETSc solver options or string matching a default set thereof
+      solver_parameters_update:
+        Dictionary of PETSc solver options used to update the default G-ADOPT options
+      J:
+        Firedrake function representing the Jacobian of the mixed Stokes system
+      constant_jacobian:
+        Boolean specifying whether the Jacobian of the system is constant
+      nullspace:
+        A `MixedVectorSpaceBasis` for the operator's kernel
+      transpose_nullspace:
+        A `MixedVectorSpaceBasis` for the kernel of the operator's transpose
+      near_nullspace:
+        A `MixedVectorSpaceBasis` for the operator's smallest eigenmodes (e.g. rigid
+        body modes)
+    """
+
+    name = "Stokes"
+
+    def __init__(
+        self,
+        solution: fd.Function,
+        approximation: BaseApproximation,
+        T: fd.Function | float = 0.0,
+        /,
+        **kwargs,
+    ) -> None:
+        super().__init__(solution, approximation, **kwargs)
+
+        self.T = T
+
+        self.eta_ind = 2
+        self.free_surface_map = {}
+        self.buoyancy_fs = [None] * len(self.solution_split)
 
     def set_free_surface_boundary(
         self, params_fs: dict[str, int | bool], bc_id: int
     ) -> Expr:
-        # Associate the free-surface index with the boundary id
+        # Set internal degrees of freedom to zero to prevent singular matrix
+        self.strong_bcs.append(
+            InteriorBC(self.solution_space[self.eta_ind], 0.0, bc_id)
+        )
+
+        normal_stress, buoyancy = self.approximation.free_surface_terms(
+            self.solution_split[1],
+            self.T,
+            self.solution_theta_split[self.eta_ind],
+            **params_fs,
+        )
+        # Associate the free-surface index with the boundary id and buoyancy term
         # Note: This assumes that ordering of the free-surface boundary conditions is
         # the same as that of free-surface functions in the mixed space.
-        self.free_surface_map[bc_id] = self.eta_ind
+        self.free_surface_map[bc_id] = [self.eta_ind, buoyancy]
```

Comment:
Not sure if a good idea to pack these two together.

Reply:
Sorry I got a bit lost trying to work out what this has changed. Why did you put them together?

Reply:
Because I wanted to avoid having to track too many attributes. Maybe it is OK like this.